File: maximum-difference-between-increasing-elements/solution2016.py
# https://leetcode.com/problems/maximum-difference-between-increasing-elements/

from typing import List


class Solution2016:

    # T: O(n), S: O(n)
    def maximumDifference(self, nums: List[int]) -> int:
        n = len(nums)
        right_maxes = [0] * n
        right_maxes[-1] = nums[-1]
        for i in range(n - 2, -1, -1):
            right_maxes[i] = max(right_maxes[i + 1], nums[i])
        max_diff = -1
        for i in range(n):
            if right_maxes[i] > nums[i]:
                max_diff = max(max_diff, right_maxes[i] - nums[i])
        return max_diff

    # T: O(n), S: O(1)
    def maximumDifferenceOpt(self, nums: List[int]) -> int:
        n = len(nums)
        right_max = nums[-1]
        max_diff = -1
        for i in range(n - 2, -1, -1):
            if nums[i] < right_max:
                max_diff = max(max_diff, right_max - nums[i])
            right_max = max(right_max, nums[i])
        return max_diff


def main() -> None:
    sol = Solution2016()
    nums1 = [7, 1, 5, 4]
    nums2 = [9, 4, 3, 2]
    nums3 = [1, 5, 2, 10]
    print(sol.maximumDifference(nums1))  # 4
    print(sol.maximumDifference(nums2))  # -1
    print(sol.maximumDifference(nums3))  # 9
    print(sol.maximumDifferenceOpt(nums1))  # 4
    print(sol.maximumDifferenceOpt(nums2))  # -1
    print(sol.maximumDifferenceOpt(nums3))  # 9


if __name__ == "__main__":
    main()

File: maximum-difference-between-increasing-elements/Solution2016.java
// https://leetcode.com/problems/maximum-difference-between-increasing-elements/

class Solution2016 {

    // T: O(n), S: O(n)
    public int maximumDifference(int[] nums) {
        int n = nums.length;
        var rightMaxes = new int[n];
        rightMaxes[n - 1] = nums[n - 1];
        for (int i = n - 2; i >= 0; i--) {
            rightMaxes[i] = Math.max(rightMaxes[i + 1], nums[i]);
        }
        int maxDiff = -1;
        for (int i = 0; i < n; i++) {
            if (rightMaxes[i] > nums[i]) {
                maxDiff = Math.max(maxDiff, rightMaxes[i] - nums[i]);
            }
        }
        return maxDiff;
    }

    // T: O(n), S: O(1)
    public int maximumDifferenceOpt(int[] nums) {
        int n = nums.length;
        int rightMax = nums[n - 1], maxDiff = -1;
        for (int i = n - 2; i >= 0; i--) {
            if (nums[i] < rightMax) {
                maxDiff = Math.max(maxDiff, rightMax - nums[i]);
            }
            rightMax = Math.max(rightMax, nums[i]);
        }
        return maxDiff;
    }

    public static void main(String[] args) {
        var sol = new Solution2016();
        var nums1 = new int[] { 7, 1, 5, 4 };
        var nums2 = new int[] { 9, 4, 3, 2 };
        var nums3 = new int[] { 1, 5, 2, 10 };
        System.out.println(sol.maximumDifferenceOpt(nums1));
        System.out.println(sol.maximumDifferenceOpt(nums2));
        System.out.println(sol.maximumDifferenceOpt(nums3));
    }

}
